### PR TITLE
Add split and collect

### PR DESCRIPTION
**Pull request by CodeSandwich**
Part of https://github.com/radicle-dev/radicle-drips-hub/issues/97, blocked by https://github.com/radicle-dev/radicle-drips-hub/pull/109.

### CODE REVIEW

File: src/ManagedDripsHub.sol [hunk 1]
```diff
@@ -78,6 +78,35 @@ abstract contract ManagedDripsHub is DripsHub, UUPSUpgradeable {
...
+        return super.split(user, assetId, currReceivers);
+    }
+
+    /// @notice Collects user's received already split funds
+    /// and transfers them out of the drips hub contract to that user's wallet.
+    /// @param user The user
+    /// @param assetId The used asset ID
+    /// @return amt The collected amount
+    function collect(address user, uint256 assetId)
+        public
```
Comment (xmxanuel) [inline src/ManagedDripsHub.sol:102]: We will restrict this method with the `msg.sender` check in the upcoming PRs, correct?

File: src/ManagedDripsHub.sol [hunk 2]
```diff
@@ -78,6 +78,35 @@ abstract contract ManagedDripsHub is DripsHub, UUPSUpgradeable {
...
     }
 
+    /// @notice Splits user's received but not split yet funds among receivers.
+    /// @param user The user
+    /// @param assetId The used asset ID
+    /// @param currReceivers The list of the user's current splits receivers.
+    /// @return collectableAmt The amount made collectable for the user
+    /// on top of what was collectable before.
+    /// @return splitAmt The amount split to the user's splits receivers
+    function split(
```
Comment (xmxanuel) [inline src/ManagedDripsHub.sol:88]: Okay, this function will stay public because we only increase the `balance.split` and not force a erc20 transfer here

**Review by xmxanuel — APPROVED** [review]
nice work,  also very modular PRs.  🚀

File: src/DripsHub.sol [hunk 3]
```diff
@@ -334,49 +373,91 @@ abstract contract DripsHub {
...
-            }
-            collected -= split;
-        }
-        emit Collected(user, assetId, collected, split);
+    /// @notice Collects user's received already split funds
+    /// and transfers them out of the drips hub contract to that user's wallet.
+    /// @param user The user
+    /// @param assetId The used asset ID
+    /// @return amt The collected amount
+    function collect(address user, uint256 assetId) public virtual returns (uint128 amt) {
```
[the remark this answers is not in the document]
Comment (CodeSandwich) [reply src/DripsHub.sol:437]: It's a good idea, could you open an issue for that feature?

File: src/DripsHub.sol [hunk 4]
```diff
@@ -237,62 +254,56 @@ abstract contract DripsHub {
...
         address user,
         uint256 assetId,
         SplitsReceiver[] memory currReceivers
-    ) public virtual returns (uint128 collected, uint128 split) {
-        (collected, split) = _collectAllInternal(user, assetId, currReceivers);
-        _transfer(user, assetId, int128(collected));
+    ) public virtual returns (uint128 collectedAmt, uint128 splitAmt) {
+        receiveDrips(user, assetId, type(uint64).max);
+        (, splitAmt) = split(user, assetId, currReceivers);
+        collectedAmt = collect(user, assetId);
```
[the remark this answers is not in the document]
Comment (CodeSandwich) [reply src/DripsHub.sol:303]: Huh, I'm really surprised, to me the relation between `collect` and `collectAll` is pretty obvious, one is a superset of the other. OTOH `withdraw` and `collect` seem unrelated. Historically we've used to call the operation of taking funds out of the contract which were supposed to be dripped a `withdrawal`, IMO that's a well suited name for an operation like this. Do you have a strong opinion on that?

File: src/DripsHub.sol [hunk 5]
```diff
@@ -334,49 +373,91 @@ abstract contract DripsHub {
...
+        balance.unsplit = 0;
+        uint32 splitsWeight = 0;
+        for (uint256 i = 0; i < currReceivers.length; i++) {
+            splitsWeight += currReceivers[i].weight;
+            uint128 currSplitAmt = uint128(
+                (uint160(collectableAmt) * splitsWeight) / TOTAL_SPLITS_WEIGHT - splitAmt
+            );
+            splitAmt += currSplitAmt;
+            address splitsReceiver = currReceivers[i].receiver;
+            splitsStates[calcUserId(splitsReceiver)].balances[assetId].unsplit += currSplitAmt;
```
[the remark this answers is not in the document]
Comment (CodeSandwich) [reply src/DripsHub.sol:416]: We MUST NOT do that, your second comment nails the problem perfectly.

File: src/ManagedDripsHub.sol (hunk 2 above)
Comment (CodeSandwich) [reply src/ManagedDripsHub.sol:88]: Correct!

File: src/DripsHub.sol (hunk 4 above)
[the remark this answers is not in the document]
Comment (xmxanuel) [reply src/DripsHub.sol:303]: In the new  low `collect` function it basically only does the erc20 transfer, so why not call the low level process `withdraw`.

For me with the current naming i would assume the `collectAll` is multiple erc20 transfers.

However the collectAll it actually `receivesDrips`, `splits it` and then forces the erc20 transfers which is called collect.

At least that was my reasoning behind it. :) No strong opinion, however.

[the remark this answers is not in the document]
Comment (CodeSandwich) [reply src/DripsHub.sol:303]: > why not call the low level process withdraw?
`withdraw` is IMO too vague in this context, it doesn't specify which funds are being withdrawn. These could be funds previously locked for dripping. With `collect` it's clear that it's something coming from other people.

From the caller perspective `collect` and `collectAll` are the same, they both put received funds in the wallet, only the second one is more expensive and yields more funds. IMO however we call these functions, convention `xxx` and `xxxAll` is worth keeping.

**Pull request merged**